File: backend/routes/schedules.py
"""Schedule Management API."""

from fastapi import APIRouter, HTTPException
from models.schemas import ScheduleCreate
from mqtt_client import mqtt_broker

router = APIRouter(prefix="/api/schedules", tags=["schedules"])

# In-memory schedule store (replace with DB in production)
_schedules: list[dict] = []
_next_id = 1


@router.get("")
def list_schedules(device_id: str = ""):
    """List schedules, optionally filtered by device."""
    result = _schedules
    if device_id:
        result = [s for s in result if device_id in s.get("device_ids", [])]
    return {"schedules": result, "total": len(result)}


@router.post("")
def create_schedule(req: ScheduleCreate):
    """Create a new schedule and push to devices."""
    global _next_id
    sched = {
        "id": _next_id,
        "title": req.title,
        "schedule_time": req.schedule_time,
        "schedule_date": req.schedule_date,
        "repeat": req.repeat,
        "alert_before_min": req.alert_before_min,
        "device_ids": req.device_ids,
        "active": True,
    }
    _schedules.append(sched)
    _next_id += 1

    # Push to each target device
    for did in req.device_ids:
        schedules_for_device = [
            {
                "id": f"sched_{s['id']:03d}",
                "title": s["title"],
                "time": s["schedule_time"],
                "date": s["schedule_date"],
                "repeat": s["repeat"],
                "alert_before_min": s["alert_before_min"],
            }
            for s in _schedules
            if did in s.get("device_ids", []) and s["active"]
        ]
        mqtt_broker.publish_schedule(did, schedules_for_device)

    return {"status": "ok", "schedule": sched}


@router.put("/{schedule_id}")
def update_schedule(schedule_id: int, req: ScheduleCreate):
    """Update a schedule."""
    sched = next((s for s in _schedules if s["id"] == schedule_id), None)
    if not sched:
        raise HTTPException(404, "Schedule not found")

    sched.update({
        "title": req.title,
        "schedule_time": req.schedule_time,
        "schedule_date": req.schedule_date,
        "repeat": req.repeat,
        "alert_before_min": req.alert_before_min,
        "device_ids": req.device_ids,
    })
    return {"status": "ok", "schedule": sched}


@router.delete("/{schedule_id}")
def delete_schedule(schedule_id: int):
    """Delete a schedule."""
    global _schedules
    _schedules = [s for s in _schedules if s["id"] != schedule_id]
    return {"status": "ok"}

File: backend/routes/schedules.js
/**
 * ClockMQTT — Schedule Routes
 * ============================
 * /api/schedules — schedule CRUD with MQTT push.
 */

const { Router } = require('express');
const scheduleStore = require('../scheduleStore');
const mqttClient = require('../mqttClient');

const router = Router();

// GET /api/schedules — list schedules, optionally filtered by device
router.get('/', (req, res) => {
  const result = scheduleStore.list(req.query.device_id || null);
  res.json(result);
});

// POST /api/schedules — create a new schedule and push to devices
router.post('/', (req, res) => {
  try {
    const schedule = scheduleStore.create(req.body);

    // Push schedule update to all assigned devices
    if (schedule.device_ids && schedule.device_ids.length > 0) {
      const allSchedules = scheduleStore.list().schedules;
      // Push per-device: each device gets all schedules targeting it
      for (const deviceId of schedule.device_ids) {
        const deviceSchedules = allSchedules.filter(s =>
          s.device_ids && s.device_ids.includes(deviceId)
        );
        mqttClient.publishSchedule(deviceId, deviceSchedules);
      }
    }

    res.json({ status: 'ok', schedule });
  } catch (e) {
    res.status(400).json({ detail: e.message });
  }
});

// PUT /api/schedules/:id — update a schedule
router.put('/:id', (req, res) => {
  try {
    const schedule = scheduleStore.update(parseInt(req.params.id, 10), req.body);
    if (!schedule) {
      return res.status(404).json({ detail: 'Schedule not found' });
    }

    // Push updated schedule list to devices
    if (schedule.device_ids && schedule.device_ids.length > 0) {
      const allSchedules = scheduleStore.list().schedules;
      for (const deviceId of schedule.device_ids) {
        const deviceSchedules = allSchedules.filter(s =>
          s.device_ids && s.device_ids.includes(deviceId)
        );
        mqttClient.publishSchedule(deviceId, deviceSchedules);
      }
    }

    res.json({ status: 'ok', schedule });
  } catch (e) {
    res.status(400).json({ detail: e.message });
  }
});

// DELETE /api/schedules/:id — delete a schedule
router.delete('/:id', (req, res) => {
  const deleted = scheduleStore.delete(parseInt(req.params.id, 10));
  if (!deleted) {
    return res.status(404).json({ detail: 'Schedule not found' });
  }
  res.json({ status: 'ok' });
});

module.exports = router;
